### PR TITLE
add management command that can set cleanliness flags

### PR DESCRIPTION
**Pull request by czue**
phase II of ownership cleanliness https://docs.google.com/document/d/12WfZLerFL832LZbMwqRAvXt82scdjDL51WZVNa31f28/edit#heading=h.gu9sjekp0u2p

@dannyroberts and @snopoke are probably most familiar with this

cc @emord

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/phone/cleanliness.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                return CleanlinessFlag(False, indexed_with_right_owner.pop())
+
+            # the only way we can get here is if an owner id spans multiple domains
+            # (and therefore has unclean indices, but not in this domain)
+            # in this case it should be clean for our domain so default to clean below
+
+    return CleanlinessFlag(True, None)
+
+
+def get_case_footprint_info(domain, owner_id):
```
Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/cleanliness.py]: @snopoke I think maybe we can use something like this to make restore faster but not 100% sure yet (haven't profiled against existing logic)

File: corehq/ex-submodules/casexml/apps/phone/models.py [hunk 2]
```diff
@@ -357,7 +357,7 @@ def __unicode__(self):
         return "%s synced on %s (%s)" % (self.user_id, self.date.date(), self.get_id)
 
 
-class OwnershipCleanliness(models.Model):
+class OwnershipCleanlinessFlag(models.Model):
```
Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: I'm doing something a bit sketchy though quite convenient here.

This database table is completely empty on all of our servers (logic was such that you would never save to it unless something was already there which was dependent on this PR for happening).

At the same time, I forgot to check in the initial migration with the original model.

By renaming the model (to a more appropriate name, i'll add) I get to add a migration file at the same time AND change the schema AND not have to worry about running any manual fake migrations on prod.

The downside is there will be an empty db table hanging around for the old model, but I'll go clean those out after this gets merged and deployed.

File: corehq/ex-submodules/casexml/apps/phone/models.py [hunk 3]
```diff
@@ -365,7 +365,7 @@ class OwnershipCleanliness(models.Model):
     We use this field to optimize restores.
     """
     domain = models.CharField(max_length=100, db_index=True)
-    owner_id = models.CharField(max_length=100, db_index=True, primary_key=True)
+    owner_id = models.CharField(max_length=100, db_index=True)
```
Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: of course our data is messy and `owner_id`s span domains. this is sometimes true for random IDs because of test/api users, and especially true of owners like `demo_user` and `commtrack-system`

File: corehq/ex-submodules/casexml/apps/phone/cleanliness.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    """
+    Get all owner ids that are assigned to cases in a domain.
+    """
+    key = ["all owner", domain]
+    submitted = CommCareCase.get_db().view('case/all_cases',
+        group_level=3,
+        startkey=key,
+        endkey=key + [{}],
+    ).all()
+    return set([row['key'][2] for row in submitted])
```
Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/phone/cleanliness.py]: How would you feel about instituting a new convention that all (new) calls to couchdb views need to go into a file called `dbaccessors.py` in the app corresponding to the design doc it's referencing (so in this case `casexml.apps.case.dbaccessors`)? The benefits of this are:

- the app itself more or less tightly controls the interface through which callers access the database
- all of these functions are in one place, making it easy to see if they can be combined or replaced, or if there are duplicates
- it's a first step towards being able to rewrite how we interface with our data—whether that's consolidating and removing couch views or changing the storage and querying backend altogether
- it checks out against with my gut as being a "nice" organizational pattern anyway

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/cleanliness.py]: i like that idea a lot. would prefer to do in a separate PR if there's nothing else blocking this, but can look into that soon

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/cleanliness.py]: (updated for just this view, but will look at other ones for another PR)

File: corehq/ex-submodules/casexml/apps/case/dbaccessors.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    """
+    Get all owner ids that are assigned to cases in a domain.
+    """
+    key = ["all owner", domain]
+    submitted = CommCareCase.get_db().view('case/all_cases',
+        group_level=3,
+        startkey=key,
+        endkey=key + [{}],
+    ).all()
+    return set([row['key'][2] for row in submitted])
```
Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/case/dbaccessors.py]: Great! Yeah that sounds good

File: corehq/ex-submodules/casexml/apps/phone/models.py (hunk 2 above)
Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: Ha nice. Seems practical

**Comment by czue** [comment]
this is :white_check_mark: from my perspective. would be nice (but not critical) to get it in before today's deploy

File: corehq/ex-submodules/casexml/apps/phone/cleanliness.py (hunk 1 above)
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/cleanliness.py]: Nice, I like the simplicity of the algorithm.

File: corehq/ex-submodules/casexml/apps/phone/models.py (hunk 2 above)
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: So after you run the migration we'll still just be updating any flags already in the DB and not adding new ones correct?

File: corehq/ex-submodules/casexml/apps/phone/models.py (hunk 3 above)
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: shucks, how lovely.

File: corehq/ex-submodules/casexml/apps/phone/models.py (hunk 2 above)
Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: correct

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: unless i run this on a particular/all domains

File: corehq/ex-submodules/casexml/apps/phone/cleanliness.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            # it wasn't in any of the open or closed IDs - it must be dirty
+            case_id_outside_footprint = cases_to_check.pop()
+            reverse_index_ids = set(get_reverse_indexed_case_ids(domain, [case_id_outside_footprint]))
+            indexed_with_right_owner = (reverse_index_ids & (footprint_info.base_ids | closed_owned_case_ids))
+            if indexed_with_right_owner:
+                return CleanlinessFlag(False, indexed_with_right_owner.pop())
+
+            # the only way we can get here is if an owner id spans multiple domains
+            # (and therefore has unclean indices, but not in this domain)
+            # in this case it should be clean for our domain so default to clean below
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/cleanliness.py]: Interesting, took me a while to get this right in my head since `footprint_info.base_ids` and `closed_owned_case_ids` could have cases from other domains.

**Comment by snopoke** [comment]
This is nice, I like the util functions, feels like we could use them in a bunch of other places too.

File: corehq/ex-submodules/casexml/apps/phone/cleanliness.py (hunk 6 above)
Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/cleanliness.py]: yeah, I think we want a view that indexes by domain AND owner, but I couldn't find that and didn't seem worth adding a new view just yet.

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/cleanliness.py]: @snopoke turns out I was totally wrong about this (since the reverse index check filters by domain). changed/fixed in https://github.com/dimagi/commcare-hq/pull/6626/
